extensions/smallrye-opentracing/deployment/pom.xml

```diff
@@ -30,6 +30,20 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-smallrye-opentracing</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.smallrye</groupId>
+            <artifactId>smallrye-fault-tolerance-tracing-propagation</artifactId>
```

Comment:
So this is what I'd like to ask about. I'm not exactly sure where to add this dependency so that it also works in user applications (it obviously works in the unit test, which is in the same module :-) ).

The intent is that this module is available when both `quarkus-smallrye-fault-tolerance` and `quarkus-smallrye-opentracing` are present.

Reply:
/cc @aloubyansky @stuartwdouglas 

I don't think we have anything for that. Not sure how we could solve this.

Reply:
Some extensions already bring other extensions transitively (e.g. I think Fault Tolerance brings Context Propagation). We can do the same here, but not sure from which side :-)

Reply:
@Ladicek by work in user applications you mean as a runtime dependency?

To this day, runtime dependencies remain in user's hands, expressed in traditional maven/gradle way. And e.g. `dependency:tree` (and the gradle equivalents) is still showing the actual runtime dependency tree. From this perspective, it's not like we don't have anything for that, it's the driving build system.
We could definitely come up with an enhancement to be able to stick other dependencies into the runtime classpath, it will break what's mentioned above though.

I think today, you'd have to either add it as a dependency of the runtime artifact or ask users to add it explicitly as an application dependency.

Reply:
> Some extensions already bring other extensions transitively (e.g. I think Fault Tolerance brings Context Propagation). We can do the same here, but not sure from which side :-)

The least popular one :)

Reply:
Hi @michalszynkiewicz @pavolloffay @loicmathieu, do you have an opinion here? Would it be better for Fault Tolerance to automatically bring OpenTracing, or the other way around?

Reply:
@Ladicek isn't a mistake (my mistake :D) that `smallrye-fault-tolerance-tracing-propagation` depends on `smallrye-fault-tolerance`?
If we could drop this dependency, we don't need to bring the whole fault tolerance with opentracing, right?

Reply:
Yea, correct. In fact, just moving the dependency on `smallrye-fault-tolerance-tracing-propagation` to the `runtime` module should do, no? It brings some unnecessary dependencies on SmallRye FT, but those should basically be ignored, because the Quarkus extension isn't present. Or is that a complete misunderstanding?

Reply:
This is actually a relatively common use case, so I think we should add something to handle it in our resolution/classloading infrasturucture.

There are lots of cases like this where you only want an integration module if two different dependencies are present, I am not 100% sure what the best way to express/handle this would be though.

Reply:
> It brings some unnecessary dependencies on SmallRye FT, but those should basically be ignored, because the Quarkus extension isn't present.

In JVM mode the JAR will be included inside the final artifact by Maven. We do nothing about it. I'm not sure we can prune some dependencies out of the build easily.

You should use one of these two Maven tricks to avoid pulling all dependencies :
- Add your dependency as optional: it's what we do for health and metrics, not sure it can be used here as you want to include this one if both are present ...
- Exclude all transitive dependencies that will be included if the other libary (fault tolerance) is included.

Then use the `Capabilities` facility to check if both FT and context-propagation are present and add the functionality inside the Quarkus build. But apparently the fonctionality is automatically added thanks to the presence of the dependency so this can be tricky.

I think the best is what @michalszynkiewicz suggest: make `smallrye-fault-tolerance-tracing-propagation` not depending on `smallrye-fault-tolerance` if possible.

Reply:
Oh, dependency exclusions should work. Basically, `quarkus-smallrye-opentracing` would depend on `smallrye-fault-tolerance-tracing-propagation` (but that dependency would exclude all transitive dependencies). It only contains 1 class and a `ServiceLoader` file for MP Context Propagation. Hmm, I wonder what would happen if this would be present, ConProp would be present, but FT would be missing... Well, since `smallrye-fault-tolerance-tracing-propagation` doesn't really depend on anything FT, we would probably do just the right thing -- have tracing context propagated. I need to test this.

(Rant: OpenTracing needs to integrate with ConProp, then we could remove this from FT entirely. We need ConProp in MP Platform!)

Reply:
OK, that works, I've updated the PR.

Reply:
I mean: I have made `smallrye-fault-tolerance-tracing-propagation` a dependency of `quarkus-smallrye-opentracing`, but all transitive dependencies are excluded. This works both when Fault Tolerance is present and missing. This is because `smallrye-fault-tolerance-tracing-propagation` in fact integrates MP OpenTracing with MP Context Propagation, and it only lives in SRye Fault Tolerance by accident.

I've tested that in an external test suite; adding tests to Quarkus would require adding 1 or 2 Maven modules, which is something I'd like to avoid.